### ISSUE TITLE
TS(TestSet): Create first profile after install service and rs-mobile

### ISSUE THREAD

**Issue by obrasero** [RetroShare/retroshare-mobile#32]
#### apks
-  https://github.com/RetroShare/retroshare-mobile/pull/30#issuecomment-893253792
- service:     RetroShare_Android_Service-alpha-arm.apk at 05.08.
- rs-mobile: app-release.apk build(avin) at 05.08.

#### Steps: Create the first profile after install
1. OK: Copied both apks to the download section of mobile
2. OK: In Download click Service and install
3. OK: in Doenload click app-release.apk and install
4. OK: in Apps find "R": click service app
  4.1.  => show with START-Button
5. OK: click START
  5.1. => shows STOP button
6. OK: in Apps find "R": click rs mobile
  6.1. => FAIL: shows entry view to create profile
    - =>      {instead it shows Can' connect...}
7. OK: in Apps find "R": click service app
  7.1 => FAIL: show still STOP
    - =>      {instead it shows START; the service got shut down as mosty - by crash or other reasoen i dont know}

**Comment by Kumaravinash9** [comment]
There is some problem with retroshare service.

**Comment by obrasero** [comment]
So i just did the above by debugging in as.
- OK:service started
- OK:Debug rs-mobile
- OK: clienz got installed on mobile and started within start window.
- OK: typed in the data to creater a first profile
- OK: pushed: create it
  1. => AS-debugger running a while through the code than a Socket-Exception appeard
  2. => checking the state of the service in mobile: stoped.
  
  {while typing the mobile did go to "sleep" and after waiking/mobile login it again the debugger did close so i couldn not check what the exception told.... i do it again an dtry to hold the mobile awaikend}

**Comment by Kumaravinash9** [comment]
If this is throwing you socket exception than it is the problem with retroshare service. because the UI part seems good to me.

**Comment by obrasero** [comment]
yes in one way of view:
- the service should not crash/throw unawaited.
but the question could be:
- is the service crashing caused by its alone only because of he has to process request
- or may it be that he crashes caused by delivered in any way invalid requests.

what do i need in my build system to build and debug the service itself?

and aside from this is the debugging with your actula test code making me slowly but definitly mor familiar with the rs-mobile and flutter-dart code. so in any way at least a win to me ;)
---

another thing:
i think the client rs-mobile should if not restarting the service by its own at least allway tell the poorly user what happens and how he can get it again into life.
but as long as the service goes down in common case its not helping as it is not realy usable.
so we will have to find out what to fix in this.
maybe i have luck and my doings give the team the right idea, or?

**Comment by obrasero** [comment]
@Kumaravinash9
ok. i listed above now the complete rquest and the reaction in the callstack upwards until user view.

**Comment by obrasero** [comment]
do you now how a naive flutter-as beginner can setup an environment to debug the service itself? to get an idea why he socket refuses?

**Comment by defnax** [comment]
maybe some one know more about this issue @G10h4ck @sehraf

**Comment by selankon** [comment]
Hey!

I saw in your exception this:

```
 exception = {SocketException} SocketException: OS Error: Connection refused, errno = 111, address = localhost, port = 50266
```

Seem that is trying to connect  to the port `50266`, usually RS service run on `9092`. But then at the POST operation you are pointing correctly the port 9092.

Some tricks:

To debug RetroShare on the phone you can use: 

```
adb logcat | grep -i retro
```

Also you can do curl calls from the computer via adb using, this can help you to debug: 

```
adb forward tcp:9091 tcp:9092
curl -u http://127.0.0.1:9091/RsJsonApi/version -vv
```

**Try to debug further and paste the results**

And about that:

> i think the client rs-mobile should, if not restarting the service by its own, at least allways tell the poore user what happens and how he can get it again into life.

We resolve it [here](https://gitlab.com/elRepo.io/elRepo.io-android/-/issues/37#note_617425981).

**Comment by obrasero** [comment]
> To debug RetroShare on the phone 
I debug it with android studio on the phone.

Ok, tomorrow i will again debug it step by step.

> We resolve it here.
As i understood, this is up to restart the downed service automatic again:OK

Are you aware or do you know that the service stay active in the background an endless time if no rs-mobile is active/installed als if you wotk wth other apps in the phone meanwhile.
If documented this with: #33

**Comment by selankon** [comment]
> I debug it with android studio on the phone.

Yes but you are debugging the flutter app, not the retroshare service, youshould know why RS service fall if its falling.


To mantain a service up, on android, you will need to run this service as a foreground application, using an android notification to mantain it up.

On elrepo.io, we bundle the retroshare service apk on the flutter app and make it runing as foreground application. I think you will find how to do it on the issue i quoted. Otherwhise ask me of how to do this process.

If you dont do that, androidOS will set the service in idle status or something similar after a while, stopping RS servic until focus.

https://developer.android.com/guide/components/foreground-services

**Comment by Kumaravinash9** [comment]
> > I debug it with android studio on the phone.
> 
> Yes but you are debugging the flutter app, not the retroshare service, youshould know why RS service fall if its falling.
> 
> To mantain a service up, on android, you will need to run this service as a foreground application, using an android notification to mantain it up.
> 
> On elrepo.io, we bundle the retroshare service apk on the flutter app and make it runing as foreground application. I think you will find how to do it on the issue i quoted. Otherwhise ask me of how to do this process.
> 
> If you dont do that, androidOS will set the service in idle status or something similar after a while, stopping RS servic until focus.
> 
> https://developer.android.com/guide/components/foreground-services

Hii @selankon ,
Actually the problem with the retroshare service.Not with the retroshare mobile UI. I have tested it many times. One more help I need from your side that can you help me  to create a new retroshare service mobile apk through retroshare repo.

**Comment by Kumaravinash9** [comment]
Could you please send me the proper steps how to compile the new retroshare service apk. Thanks

**Comment by obrasero** [comment]
> Could you please send me the proper steps how to compile the new retroshare service apk. Thanks

Yes, as is asked above also "do you now how ... can setup an environment to debug the service itself?" that would be very good/helpfull.

**Comment by selankon** [comment]
> > Could you please send me the proper steps how to compile the new retroshare service apk. Thanks
> 
> Yes, as is asked above also "do you now how ... can setup an environment to debug the service itself?" that would be very good/helpfull.

Hi,

To debug service itself, as i suggested, use abive, use :

```
adb logcat | grep -i retro
```

And you can do `curl` to check responses.

> Could you please send me the proper steps how to compile the new retroshare service apk

@G10h4ck do the compilation. Maybe can help.

In the following days I could do a step by step guidw of how to integrate retroshare apk inside a flutter apk and execute it as foreground service.

**Comment by obrasero** [comment]
> @selankon 
> To debug service itself, as i suggested, use abive, use :
>> adb logcat | grep -i retro

Ahh...,
no the android-studio newbe got :)
I di have to open an cmd in  the dir of the andoid palttform tools
OK. i got now the trace
so will now look through.... be pacient with the slow newbe to get your hint ;)

**Comment by obrasero** [comment]
OK i now did a new testrun with the rs-mobile from avins master:
Actual a nome more socket exceptions pop up like the last time.
now an
"Unhandled Exception: Connection closed before full header was received"
always appears on first account creation.

now i attach the adb trace:
---

[trace.txt](https://github.com/RetroShare/retroshare-mobile/files/6949709/trace.txt)

**Comment by Kumaravinash9** [comment]
https://stackoverflow.com/questions/55879550/how-to-fix-httpexception-connection-closed-before-full-header-was-received
@obrasero could you please try this

**Comment by obrasero** [comment]
Thanks for the hint,
but sadly flutter clean doesnt fix it.
and it happen to my on physical device in release and debug builds.

**Comment by obrasero** [comment]
But i checked out your new master and build tested with it:
Don't know why but with this the "before full header" did not happen in profile/account-creation, so now i was able to create an account with just before manually restarting the service. 
but while playing around it happened at other actions as well as the socket exception. but always manually starting the service before next action i was at least able to created identities.

**Comment by Kumaravinash9** [comment]
More commonly, it is caused by writing to a connection that the other end has already closed normally. In other words an application protocol error. It can also be caused by closing a socket when there is unread data in the socket receive buffe

**Comment by Kumaravinash9** [comment]
> > > Could you please send me the proper steps how to compile the new retroshare service apk. Thanks
> > 
> > 
> > Yes, as is asked above also "do you now how ... can setup an environment to debug the service itself?" that would be very good/helpfull.
> 
> Hi,
> 
> To debug service itself, as i suggested, use abive, use :
> 
> ```
> adb logcat | grep -i retro
> ```
> 
> And you can do `curl` to check responses.
> 
> > Could you please send me the proper steps how to compile the new retroshare service apk
> 
> @G10h4ck do the compilation. Maybe can help.
> 
> In the following days I could do a step by step guidw of how to integrate retroshare apk inside a flutter apk and execute it as foreground service.
Hi @selankon ,
Could you please help me to do this ?

**Comment by selankon** [comment]
@Kumaravinash9 I wrote this guide of how to bundle RS service as Android Foreground service

https://gitlab.com/elRepo.io/elRepo.io-android/-/issues/43

Probably I forgot stuff, write me a comment there if need to ask

**Comment by Kumaravinash9** [comment]
Thanks @selankon. And can you help me to debug this login problem. Actually from my side i pointed port at 9092 but still comes socket exception.

**Comment by selankon** [comment]
Please @Kumaravinash9 provide related information:

* What code are you building?
* RS Service and flutter trace and errors
* Is `curl` working? If not, use `-vv` version and paste the trace here. Remember you will need to do a port forwarding from adb as show above. 

Maybe Android bring down the RetroShare service into a idle state?  Please provide as much information as you can when asking for a solution.

**Comment by obrasero** [comment]
> Is curl working? 

Yes, at least at my environment:

    curl http://127.0.0.1:9091/RsJsonApi/version -v
    *   Trying 127.0.0.1...
    * TCP_NODELAY set
    * Connected to 127.0.0.1 (127.0.0.1) port 9091 (#0)
    > GET /RsJsonApi/version HTTP/1.1
    > Host: 127.0.0.1:9091
    > User-Agent: curl/7.55.1
    > Accept: */*
    >
    < HTTP/1.1 200 OK
    < Access-Control-Allow-Headers: Authorization,DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range
    < Access-Control-Allow-Methods: GET, POST, OPTIONS
    < Access-Control-Allow-Origin: *
    < Access-Control-Expose-Headers: Content-Length,Content-Range
    < Connection: close
    < Content-Length: 116
    < Content-Type: application/json
    <
    {
        "major": 0,
        "minor": 6,
        "mini": 6,
        "extra": "-38-g25f58bc10",
        "human": "0.6.6-38-g25f58bc10"
    }* Closing connection 0

**Comment by obrasero** [comment]
>  @selankon
> RS Service and flutter trace and errors

Appended ad trace with socket error...
[trace socketexception.txt](https://github.com/RetroShare/retroshare-mobile/files/6957607/trace.socketexception.txt)

**Comment by obrasero** [comment]
..and here a trace with "Connection closed before full header was received"
[trace debug after f clean.txt](https://github.com/RetroShare/retroshare-mobile/files/6957634/trace.debug.after.f.clean.txt)

**Comment by selankon** [comment]
M... maybe activity manager is killing RetroShare service

```
08-08 11:39:50.759   694  1980 I ActivityManager: Process org.retroshare.service:rs (pid 28842) has died: fore SVC 
```

@Kumaravinash9 are you trying to bundle both apks as shown in the [guide](https://gitlab.com/elRepo.io/elRepo.io-android/-/issues/43)? Maybe this fix the problem?

Could you try to do a curl call after a socket exception is thrown?

**Comment by Kumaravinash9** [comment]
> M... maybe activity manager is killing RetroShare service
> 
> ```
> 08-08 11:39:50.759   694  1980 I ActivityManager: Process org.retroshare.service:rs (pid 28842) has died: fore SVC 
> ```
> 
> @Kumaravinash9 are you trying to bundle both apks as shown in the [guide](https://gitlab.com/elRepo.io/elRepo.io-android/-/issues/43)? Maybe this fix the problem?
> 
> Could you try to do a curl call after a socket exception is thrown?

Yes working on it. will integrate it by tomorrow.

**Comment by obrasero** [comment]
Update of the test to #39:

Still not working on my phone,
also the phenomenology changed.
see: https://github.com/RetroShare/retroshare-mobile/pull/29#issuecomment-897859711

Instead of socket of not finished header async return now the exception is:
"Exception: HttpException: Exception: something went wrong!"

**Comment by obrasero** [comment]
Test repeated with an emulated device:
Results is the same: exception and hanging in Splash: "get locations..."
